src/Controller/MakefileTraceToJobTree.ts

```diff
@@ -0,0 +1,215 @@
+ * @returns a list of target-lines and their associated commands.
+ */
+function extractTargetLinesAndCommands(
+	traceLines: string[],
+): TargetLineAndCommands[] {
+	const targetLinesWithCommands: TargetLineAndCommands[] = []
+
+	for (const line of traceLines) {
+		if (
+			targetLineRegexWithPrereqs.test(line) ||
+			targetLineRegexWithoutPrereqs.test(line)
+		) {
+			targetLinesWithCommands.push({ targetLine: line, commands: [] })
+		} else {
+			assert(
+				targetLinesWithCommands.length > 0,
+				`Makefile command supplied before any targets:\n${line}`,
+			)
+			targetLinesWithCommands[
+				targetLinesWithCommands.length - 1
+			]?.commands.push(line)
+		}
+	}
+
+	return targetLinesWithCommands
+}
+
+/**
+ * Creates a DAG tree of Jobs from a list of targets and their associated commands.
+ *
+ * @param targetLinesWithCommands - the targets and their associated commands.
+ * @returns the tree's root jobs.
+ */
+function constructDAGFromTargetsAndCommands(
+	targetLinesWithCommands: TargetLineAndCommands[],
+): Set<Job> {
+	// We need to know whether a Job has dependents so we can isolate the root jobs later.
+	const targetToJAD = new Map<string, JobAndDependency>()
+
+	// Construct DAG.
+	for (const { targetLine, commands } of targetLinesWithCommands) {
+		// Extract target and prerequisites from target line.
+		const { target, prerequisiteJobTargets } = extractInfoFromTargetLine(
+			targetLine,
+		)
+
+		// Find all prerequisite jobs. If no Job has some target, then the target is (hopefully) a preexisting file and we can ignore it. Also marks prerequisite Jobs as having dependents.
+		const prerequisiteJobs: Set<Job> = new Set()
+		const prerequisiteFiles: Set<string> = new Set()
+		for (const prerequisiteJobTarget of prerequisiteJobTargets) {
+			const prerequisiteJobAndDependency = targetToJAD.get(
+				prerequisiteJobTarget,
+			)
+
+			if (prerequisiteJobAndDependency) {
+				prerequisiteJobs.add(prerequisiteJobAndDependency.job)
+				prerequisiteJobAndDependency.hasDependent = true
+			} else {
+				prerequisiteFiles.add(prerequisiteJobTarget)
+			}
+		}
+
+		// We should not be seeing the same target twice.
+		assert(
+			!targetToJAD.has(target),
+			`Target "${target}" exists twice in the provided Makefile trace.`,
+		)
+		targetToJAD.set(target, {
+			job: new NormalJob({
+				target,
+				commands,
+				prerequisiteJobs,
+				prerequisiteFiles,
+			}),
+			hasDependent: false,
+		})
+	}
+
+	// Get root jobs.
+	const rootJobs = new Set<Job>()
+	for (const jobAndDependency of targetToJAD.values()) {
+		if (!jobAndDependency.hasDependent) {
+			rootJobs.add(jobAndDependency.job)
+		}
+	}
+
+	return rootJobs
+}
+
+/**
+ * Filters an array of trace-lines.
+ * Removes lines that are unnecessary to the DAG constructor.
+ *
+ * @param traceLines - an array of strings, each of which is a line from a trace.
+ * @returns the filtered array of strings.
+ */
+function filterTraceLines(traceLines: string[]): string[] {
```

Comment:
How do you know this function is exhaustive? I.e. that there isn't some rarely used output from make that will break this if it appears? If possible, use a whitelist approach instead of a blacklist (but idk anything about the format, so that might not be possible).

Reply:
How do I check whether a line is a command or not?

Maybe I could throw an error if there are any lines starting with `make: ` remaining at the end of the function, but I'm not sure that a command couldn't start with that.

Reply:
It should be fine. I took a look at GNU Make's source code (a relevant output snippet is attached below), and it seems like we're testing for all possibilities. 


```
  /* Trace the build.
     Use message here so that changes to working directories are logged.  */
  if (trace_flag)
    {
      char *newer = allocated_variable_expand_for_file ("$?", c->file);
      const char *nm;

      if (! cmds->fileinfo.filenm)
        nm = _("<builtin>");
      else
        {
          char *n = alloca (strlen (cmds->fileinfo.filenm) + 1 + 11 + 1);
          sprintf (n, "%s:%lu", cmds->fileinfo.filenm, cmds->fileinfo.lineno);
          nm = n;
        }

      if (newer[0] == '\0')
        OSS (message, 0,
             _("%s: target '%s' does not exist"), nm, c->file->name);
      else
        OSSS (message, 0,
              _("%s: update target '%s' due to: %s"), nm, c->file->name, newer);

      free (newer);
    }
```

Reply:
Where do the `make: Entering directory` strings come from?